# Big Data Wrangling With Google Books Ngrams
## Part 1: Introduction to Dataset

**BrainStation**

## Introduction

In this notebook, we will look into the Google Ngrams dataset. It was created by Google's research team by analyzing all of the content in Google Books - these digitized texts represent approximately 4% of all books ever printed, and span a time period from the 1800s into the 2000s.

We will filter and reduce data down to a manageable size, so that we can do some analysis locally on our machine after extracting data from the Cloud and processing it using Big Data tools.

***

## Table of Contents <a class="anchor" id="toc"></a>

- [The Dataset](#the-dataset)
- [Creating a new dataframe](#new-dataframe)
- [Conclusion](#conclusion)

***

## The Dataset <a class="anchor" id="the-dataset"></a>

Let's do some configurations and import the relevant packages to analyze the data.

In [1]:
spark

Starting Spark application


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [2]:
%%configure -f
{
    "conf": {
        "spark.pyspark.python": "python3",
        "spark.pyspark.virtualenv.enabled": "true",
        "spark.pyspark.virtualenv.type":"native",
        "spark.pyspark.virtualenv.bin.path":"/usr/bin/virtualenv"
    }
}

Starting Spark application


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


In [3]:
from pyspark.sql.functions import col

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Now, let's read in the data and take a first look at the dataset.

In [4]:
# Let's import the file
df = spark.read.csv('s3://brainstation-dsft/eng_1M_1gram.csv', header=True)
df

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

DataFrame[token: string, year: string, frequency: string, pages: string, books: string]

In [5]:
# Shows the first 10 rows
df.show(10, vertical=True)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

-RECORD 0--------------
 token     | inGermany 
 year      | 1927      
 frequency | 2         
 pages     | 2         
 books     | 2         
-RECORD 1--------------
 token     | inGermany 
 year      | 1929      
 frequency | 1         
 pages     | 1         
 books     | 1         
-RECORD 2--------------
 token     | inGermany 
 year      | 1930      
 frequency | 1         
 pages     | 1         
 books     | 1         
-RECORD 3--------------
 token     | inGermany 
 year      | 1933      
 frequency | 1         
 pages     | 1         
 books     | 1         
-RECORD 4--------------
 token     | inGermany 
 year      | 1934      
 frequency | 1         
 pages     | 1         
 books     | 1         
-RECORD 5--------------
 token     | inGermany 
 year      | 1935      
 frequency | 1         
 pages     | 1         
 books     | 1         
-RECORD 6--------------
 token     | inGermany 
 year      | 1938      
 frequency | 5         
 pages     | 5         
 books     | 5  

In [6]:
# Counts the number of row in the data set
print(f'There are {df.count()} rows and {len(df.columns)} columns in this dataset.')

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

There are 261823225 rows and 5 columns in this dataset.

We can print out the schema and data types using `printSchema()` to see what we're working with.

In [7]:
df.printSchema()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- token: string (nullable = true)
 |-- year: string (nullable = true)
 |-- frequency: string (nullable = true)
 |-- pages: string (nullable = true)
 |-- books: string (nullable = true)

***
## Creating a new dataframe <a class="anchor" id="new-dataframe"></a>

For this notebook, we want to create a new dataframe to include only the rows where token is "data".

In [8]:
# Filters df to only include rows where token = 'data'
filtered_df = df.filter(col('token') == 'data')
filtered_df

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

DataFrame[token: string, year: string, frequency: string, pages: string, books: string]

Now, we want to take a look at the new dataframe we made.

In [11]:
filtered_df.show(10, vertical = True)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

-RECORD 0---------
 token     | data 
 year      | 1584 
 frequency | 16   
 pages     | 14   
 books     | 1    
-RECORD 1---------
 token     | data 
 year      | 1614 
 frequency | 3    
 pages     | 2    
 books     | 1    
-RECORD 2---------
 token     | data 
 year      | 1627 
 frequency | 1    
 pages     | 1    
 books     | 1    
-RECORD 3---------
 token     | data 
 year      | 1631 
 frequency | 22   
 pages     | 18   
 books     | 1    
-RECORD 4---------
 token     | data 
 year      | 1637 
 frequency | 1    
 pages     | 1    
 books     | 1    
-RECORD 5---------
 token     | data 
 year      | 1638 
 frequency | 2    
 pages     | 2    
 books     | 1    
-RECORD 6---------
 token     | data 
 year      | 1640 
 frequency | 1    
 pages     | 1    
 books     | 1    
-RECORD 7---------
 token     | data 
 year      | 1642 
 frequency | 1    
 pages     | 1    
 books     | 1    
-RECORD 8---------
 token     | data 
 year      | 1644 
 frequency | 4    
 pages     |

In [9]:
filtered_df.printSchema()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- token: string (nullable = true)
 |-- year: string (nullable = true)
 |-- frequency: string (nullable = true)
 |-- pages: string (nullable = true)
 |-- books: string (nullable = true)

In [10]:
# Counts the number of row in the data set
print(f'There are {filtered_df.count()} rows and {len(filtered_df.columns)} columns in this dataset.')

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

There are 316 rows and 5 columns in this dataset.

In [12]:
# Exporting new dataframe to csv file
filtered_df.write.csv('filtered_df.csv', header = True)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

An error was encountered:
path hdfs://ip-172-31-3-210.us-east-2.compute.internal:8020/user/livy/filtered_df.csv already exists.
Traceback (most recent call last):
  File "/mnt/yarn/usercache/livy/appcache/application_1686153419480_0010/container_1686153419480_0010_01_000001/pyspark.zip/pyspark/sql/readwriter.py", line 1240, in csv
    self._jwrite.csv(path)
  File "/mnt/yarn/usercache/livy/appcache/application_1686153419480_0010/container_1686153419480_0010_01_000001/py4j-0.10.9.5-src.zip/py4j/java_gateway.py", line 1322, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/mnt/yarn/usercache/livy/appcache/application_1686153419480_0010/container_1686153419480_0010_01_000001/pyspark.zip/pyspark/sql/utils.py", line 196, in deco
    raise converted from None
pyspark.sql.utils.AnalysisException: path hdfs://ip-172-31-3-210.us-east-2.compute.internal:8020/user/livy/filtered_df.csv already exists.



***

## Conclusion <a class="anchor" id="conclusion"></a>

We were able to read in the Google Ngrams dataset. It has 261,823,225 rows and 5 columns. We then created a new dataframe where we filtered and reduced data down to a manageable size by only including rows where the value for `token` is `data`. The new dataframe has 316 rows and 5 columns. We saved it as a csv file in the cloud so that we can extract it and do analysis on our local machines.

[[Back to top]](#toc)